### PR TITLE
[Merged by Bors] - feat: `propose`, a forwards-reasoning analogue of `library_search`.

### PR DESCRIPTION
**Pull request by kim-em**
* `propose using a, b, c` tries to find a lemma
which makes use of each of the local hypotheses `a, b, c`,
and reports any results via trace messages.
* `propose : h using a, b, c` only returns lemmas whose type matches `h` (which may contain `_`).
* `propose! using a, b, c` will also call `have` to add results to the local goal state.

`propose` should not be left in proofs; it is a search tool, like `library_search`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by Parcly-Taxel — APPROVED** [review]

File: Mathlib/Tactic/SolveByElim.lean [hunk 1]
```diff
@@ -108,7 +108,28 @@ def mainGoalProc (cfg : Config := {}) (proc : MVarId → MetaM (List MVarId)) :
 
 /-- Create or modify a `Config` which calls `intro` on each goal before applying lemmas. -/
 def intros (cfg : Config := {}) : Config :=
-  mainGoalProc cfg fun g => do pure [(← g.intro1P).2]
+  cfg.mainGoalProc fun g => do pure [(← g.intro1P).2]
+
+/-- Attempt typeclass inference on each goal, before applying lemmas. -/
```
Comment (fpvandoorn) [inline Mathlib/Tactic/SolveByElim.lean:113]: Looking at the doc of `mainGoalProc` this seems to only do this on the main goal?

File: test/propose.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import Std.Data.List.Basic
+import Mathlib.Tactic.Propose
+import Mathlib.Tactic.GuardHypNums
+import Mathlib.Algebra.Associated
+
+theorem foo (L M : List α) (w : L.Disjoint M) (m : a ∈ L) : a ∉ M := fun h => w m h
+
+example (K L M : List α) (w : L.Disjoint M) (m : K ⊆ L) : True := by
+  propose using w
```
Comment (fpvandoorn) [inline test/propose.lean:9]: here and below: assuming that you want these tactics to give deterministic answers, could you add tests using [`Std.Tactic.GuardExpr.guardHyp`](https://leanprover-community.github.io/mathlib4_docs/Std/Tactic/GuardExpr.html#Std.Tactic.GuardExpr.guardHyp) to ensure that `propose` is doing the right thing?

File: Mathlib/Tactic/Propose.lean [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  let goal ← getMainGoal
+  goal.withContext do
+    let required ← terms.mapM (elabTerm · none)
+    let type ← match type with
+    | some stx => elabTermWithHoles stx none (← getMainTag) true <&> (·.1)
+    | none => mkFreshTypeMVar
+    let proposals ← propose (← proposeLemmas.get) type required
+    if proposals.isEmpty then
+      throwError "propose could not find any lemmas using the given hypotheses"
+    for p in proposals.toList.take 10 do
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Propose.lean:117]: `propose` doesn't do any reordering of the `candidates`. Would it makes sense to have `propose` stop searching once it hits `10` (configurable) results?

**Review by fpvandoorn — COMMENTED** [review]
This looks really useful! Thanks!

Some comments, but LGTM

bors d+

File: Mathlib/Tactic/SolveByElim.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Tactic/SolveByElim.lean:113]: I've added a comment explaining.

**Pull request closed without merging**